compiler/modules/front/tests/testParser.bal

```diff
@@ -42,6 +42,63 @@ function testParser(string k, string rule, string subject, string expected) retu
     test:assertEquals(actual, expected, "wrong ast");
 }
 
+@test:Config {
+    dataProvider: sourceFragments
+}
+function testTokenizer(string k, string src) returns error? {
+    int[] lines = findLineFeeds(src);
+    Tokenizer tok = new (src);
+
+    err:Syntax|Token? t = advance(tok, k, src);
+    while t is Token {
+        err:Position pos = tok.currentPos();
+        string tStr = tokenToString(t);
+        // XXX change below after making indexInLine 1-indexed
+        int tStart = lines[pos.lineNumber - 1] + pos.indexInLine;
+        string srcAtPos = src.substring(tStart, tStart + tStr.length());
+        test:assertEquals(srcAtPos, tStr);
+        t = advance(tok, k, src);
+    }
+    if k.includes("E") {
+        test:assertTrue(t is err:Syntax, "an error expected for '" + src + "'");
+    }
+}
+
+function advance(Tokenizer tok, string k, string subject) returns err:Syntax|Token? {
+    err:Syntax? e = tok.advance();
+    if e is err:Syntax {
+        if k.includes("E") {
+            return e;
+        }
+        else {
+            panic error("tokenizer error for '" + subject + "'", e);
+        }
+    }
+    return tok.current();
+}
+
+function tokenToString(Token t) returns string {
+    if t is VariableLengthToken {
+        return t[1];
```

Comment:
This won't work for escapes in strings: in this case `t[1]` is the string after escapes have been processed. In general, you cannot recover the original string from the processed string, because you can have multiple escapes with the same meaning (e.g. `\n` and `\u{A}`). So in this case, all you can do is check the source token is longer than the processed token and that the source token starts with `"`.